app/main.py

```diff
@@ -1,2 +1,68 @@
+from typing import Any, Optional
+from dataclasses import dataclass
+
+
+@dataclass
+class Node:
+    key: Any
+    value: Any
+    _next: Optional["Node"] = None
+
+
 class Dictionary:
-    pass
+    def __init__(self) -> None:
+        self.capacity: int = 8
+        self.hash_table: list[Optional[Node]] = [None] * self.capacity
+        self.length: int = 0
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        index = hash(key) % self.capacity
+        node = self.hash_table[index]
+
+        while node:
+            if node.key == key:
+                node.value = value
+                return
+            node = node._next
+
+        new_node = Node(key=key, value=value, _next=self.hash_table[index])
+        self.hash_table[index] = new_node
+        self.length += 1
+
+        if self.length / self.capacity > 0.75:
+            self._resize()
+
+    def __contains__(self, key: Any) -> bool:
+        index = hash(key) % self.capacity
+        node = self.hash_table[index]
+        while node:
+            if node.key == key:
+                return True
+            node = node._next
+        return False
+
+    def __getitem__(self, key: Any) -> Any:
+        if key not in self:
+            raise KeyError("Key not found in dictionary")
```

Comment:
The `__getitem__` method raises a `KeyError` if the key is not found, which is correct. However, the check `if key not in self:` is inefficient because it traverses the linked list twice: once in `__contains__` and again in the subsequent loop. Consider combining the logic to traverse the list only once to improve efficiency.

Reply:
AI is right, rework this to not use `if key not in self`